### PR TITLE
feat: Add zip method

### PR DESCRIPTION
**Pull request by tianfeng92**
# One-line summary

Add zip method: this method receives source directory and destination file and zips file accordingly.

### CODE REVIEW

File: src/zip.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  try {
+    const stats = fs.statSync(source);
+    if (!stats.isDirectory()) {
+      throw new Error('Invalid source folder: the source must be a directory.');
+    }
+  } catch (err) {
+    if (errCode(err) === 'ENOENT') {
+      throw new Error('Invalid source folder: not exist.');
+    }
+    throw new Error(`Failed to access source folder: ${err}`);
```
Comment (alexplischke) [inline src/zip.ts:25]: (optional) Do we need this distinction? Or is the `ENOENT` error case's message not good enough that we need our own? Happy to keep it as is, just wanted to know for my own knowledge 🙏

File: src/zip.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  specifiedFolder: string,
+): boolean {
+  // Resolve absolute paths.
+  const absoluteTargetFolder = path.resolve(targetFolder);
+  const absoluteSpecifiedFolder = path.resolve(specifiedFolder);
+
+  // Check if the target folder is the same as the specified folder (not outside).
+  if (absoluteTargetFolder === absoluteSpecifiedFolder) {
+    return false;
+  }
```
Comment (alexplischke) [inline src/zip.ts:59]: True, but does that mean we'd allow our user to _potentially_ archive the entire workspace? Or is something else preventing that somewhere?

File: src/zip.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+ * Compresses the specified source into a zip file at the destination path.
+ *
+ * @param workspace The user workspace directory.
+ * @param source The path of the directory or file to be compressed.
+ * @param dest The path where the output zip file should be saved.
+ */
+export function zip(workspace: string, source: string, dest: string) {
+  try {
+    validate(workspace, source, dest);
+    const command = getCommand(source, dest);
```
Comment (alexplischke) [inline src/zip.ts:110]: Wouldn't it be better if `getCommand` threw an error if the platform is not supported? Logically speaking, the zip function will simply do nothing, and does so silently, if for whatever reason `getCommand` returns an empty string. Isn't it better we knew it failed?

File: src/zip.ts (hunk 1 above)
Comment (tianfeng92) [reply src/zip.ts:25]: I added it just for any unknown error here. But I'm also fine to remove it. From my experiments, I can't get the error from this line 😄

File: src/zip.ts (hunk 2 above)
Comment (tianfeng92) [reply src/zip.ts:59]: hmm... make sense. Let me remove it to prevent that user behavior.

File: src/zip.ts (hunk 3 above)
Comment (tianfeng92) [reply src/zip.ts:110]: 👍

File: src/zip.ts [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ * Generates a platform-specific command string for compressing files into a zip archive.
+ *
+ * On macOS, it constructs a shell command using the `zip` utility with options to
+ * recursively zip the content, preserve symlinks, and operate quietly.
+ *
+ * On Windows, it constructs a PowerShell command using `Compress-Archive` with options to
+ * specify the source and destination paths directly, and the `-Force` option to overwrite
+ * any existing destination file.
+ *
+ * For other operating systems, throw an error to indicates an unsupported platform.
```
Comment (alexplischke) [inline src/zip.ts:81]:
```suggestion
 * For other operating systems, throw an error to indicate an unsupported platform.
```

File: src/zip.ts [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ * Such actions are prohibited, thus leading to a return value of true.
+ *
+ * @param {string} targetFolder The path to the target folder.
+ * @param {string} specifiedFolder The path to the specified folder.
+ * @returns {boolean} Returns true if the target folder is outside of the specified folder, false otherwise.
+ */
+export function isFolderOutside(
+  targetFolder: string,
+  specifiedFolder: string,
+): boolean {
```
Comment (alexplischke) [inline src/zip.ts:55]: I find the code a bit hard to read when they refer to the vars `targetFolder` and `specifiedFolder`. The names are not conclusive enough to let me understand what is what without looking up the description again and again. How about `rootFolder` and `subFolder`?

Comment (tianfeng92) [reply src/zip.ts:55]: 👍 Better!

File: src/zip.ts [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+function getCommand(source: string, dest: string): string {
+  const osPlatform = platform();
+
+  switch (osPlatform) {
+    case 'darwin':
+      return `zip -ryq "${dest}" "${source}"`;
+    case 'win32':
+      return `Compress-Archive -Path ${source} -DestinationPath ${dest} -Force`;
+    default:
+      // Throw an error if the operating system is not supported
```
Comment (alexplischke) [inline src/zip.ts:94]: That's clear from the code and the error message itself 😅 
```suggestion
```

**Review by alexplischke — APPROVED** [review]

File: src/zip.ts [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+ * @param dest The path where the output zip file should be saved.
+ */
+export function zip(workspace: string, source: string, dest: string) {
+  try {
+    validate(workspace, source, dest);
+    execSync(getCommand(source, dest));
+  } catch (error) {
+    console.error(
+      `Zip file creation failed for destination: "${dest}", source: "${source}". Error: ${error}.`,
+    );
```
[the remark this answers is not in the document]
Comment (tianfeng92) [reply src/zip.ts:112]: I agree that an error thrown by a low-level function may be unexpected. However, given that the zipping behavior is determined by user settings in the sauce configuration, such an error could be understandable and helpful.
It allows users to debug and modify their configuration as necessary. This is different from generating JUnit or Sauce reports, where a failure could be misinterpreted as a test failure, as those operations are not directly controlled by the user.

[the remark this answers is not in the document]
Comment (alexplischke) [reply src/zip.ts:112]: @tianfeng92 Mike is talking about removing this log line and instead throwing the error all the way back to the caller (the runner in that case). The caller (runner) is then the one to handle (i.e. log) the error.

[the remark this answers is not in the document]
Comment (alexplischke) [reply src/zip.ts:112]: > Oops, added this comment, but forgot to submit the review. Fine to disregard, but still wanted to bring it up.

Now that you bring this up, I'd personally prefer that approach. Generally I'm in favor of keeping user facing messaging out of low level libs. Having user facing messages in low level libs, prevents the lib from being used more broadly and robs the caller of controlling what the user should see.

[the remark this answers is not in the document]
Comment (tianfeng92) [reply src/zip.ts:112]: Make sense. Let me draft a PR to address it.

**Pull request merged**